node/src/blockchain/blockchain_interface.rs

```diff
@@ -275,40 +307,51 @@ where
-                                            block_number: u64::try_from(block_number)
-                                                .expect("Internal Error"),
-                                            from: Wallet::from(log.topics[1]),
-                                            gwei_amount,
-                                        })
-                                    }
                                     None => None,
+                                    Some(block_number) => to_gwei(U256::from(
+                                        log.data.0.as_slice(),
+                                    ))
+                                    .map(|gwei_amount| BlockchainTransaction {
+                                        block_number: block_number.as_u64(),
+                                        from: Wallet::from(log.topics[1]),
+                                        gwei_amount,
+                                    }),
                                 })
                                 .collect();
                             debug!(logger, "Retrieved transactions: {:?}", transactions);
                             // Get the largest transaction block number, unless there are no
-                            // transactions, in which case use start_block.
-                            let last_transaction_block =
-                                transactions.iter().fold(start_block, |so_far, elem| {
-                                    if elem.block_number > so_far {
-                                        elem.block_number
-                                    } else {
-                                        so_far
-                                    }
-                                });
+                            // transactions, in which case use end_block, unless get_latest_block()
+                            // was not successful.
+                            let transaction_max_block_number = if transactions.is_empty() {
+                                match end_block {
+                                    BlockNumber::Number(U64([eb])) => eb,
+                                    _ => response_block_number,
+                                }
+                            } else {
+                                transactions
+                                    .iter()
+                                    .fold(response_block_number, |so_far, elem| {
+                                        if elem.block_number > so_far {
+                                            elem.block_number
+                                        } else {
+                                            so_far
+                                        }
+                                    })
+                            };
                             Ok(RetrievedBlockchainTransactions {
-                                new_start_block: last_transaction_block + 1,
+                                new_start_block: transaction_max_block_number,
```

Comment:
Also, as long as our request wasn't rejected by the blockchain, don't we want to use the current head block + 1 as the new start block in all cases? Suppose we scan 1000 blocks and find a few dozen transactions, with the last one in the 800th block we scan. If that happens, we've already scanned the last 200 blocks in that 1000-block range, and we know there are no transactions there. Why would we want to scan them again next time? Using the current head block in all cases (except where the request fails) might even make things simpler.

Reply:
Rescanning blocks would matter less if we change the accountant to keep track of transaction hashes so it knows if it's seen a transaction before. However, the case I'm trying to avoid is the error we receive if we exceed the bounds of `BlockNumber::Latest` meaning calling web3 with the literal block number result from a call with `BlockNumber::Latest` and adding 1 will generate an error if the block number doesn't exist yet. There's a short time period where that may be true. It could be a longer time period if the blockchain service provider is having an issue and the block number isn't increasing. I can try to refine this a bit and confirm we avoid rescanning blocks.

Reply:
Would storing the older transactions be safe? I've always thought the answer is no and so that's why we've always tried to find a way not to be obligated to do that. 

But maybe there just hasn't been a good enough discussion about this problem. In the end, all transactions are stored on the blockchain anyway. But they would have to first figure out what is our consuming wallet. Which should be too hard for them.

In other words, yes, it is a security issue to have an unprotected storage of previous transaction in your machine because they could enable uncovering your identity by the repeated consuming wallet in them. So we would probably have to encrypt these entries in the storage. Would it be still workable? Probably not if there were plenty of them.  What do you think would it be sufficient if we stored only the hash of the last transaction we saw? Could we use it to double-check if we hold the right start block?

I don't know. I didn't think much about these cases.